File: leecode/301-400/331-验证二叉树的前序序列化.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        degree = 1
        for node in preorder.split(','):
            if degree == 0:
                return False
            if node == '#':
                degree -= 1
            else:
                degree += 1
        return degree == 0

File: leecode/1-500/301-400/331-验证二叉树的前序序列化.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        int degree = 1;
        int n = preorder.length();
        for (int i = 0; i < n; ++i) {
            if (preorder[i] == ',') continue;
            if (degree == 0) return false;
            if (preorder[i] == '#')
                degree -= 1;
            else {
                i ++;
                while(i < n && preorder[i] != ',' && preorder[i] != '#')
                    i ++;
                degree += 1;
                i --;
            }
        }
        return degree == 0;
    }
};